src/txn/txn.c

```diff
@@ -242,6 +242,13 @@ __wt_txn_get_snapshot(WT_SESSION_IMPL *session)
     WT_ASSERT(session, prev_oldest_id == txn_global->oldest_id);
     txn_state->pinned_id = pinned_id;
 
+    /*
+     * In general, other threads should be only be reading data exposed in the shared transaction
+     * state. There are a few instances where we need to read the internal transaction data in a
+     * controlled manner so we need to leave an escape hatch here to allow this.
+     */
+    txn_state->internal_txn = txn;
```

Comment:
Circling back to this field -- if we need this field, doesn't that mean the approach doesn't work?

Reply:
My point of view is just that the stuff in `WT_TXN_SHARED` can be read safely and easily without any caveats. Whereas reading directly from the `internal_txn` is a "do at your own risk" sort of thing and each usage of this member should be looked at carefully before going in.

The only usage at the moment is `__txn_get_published_timestamp` which I think is legitimate. Do you agree?

Reply:
@michaelcahill, would you please do a review?

Reply:
I've pushed a change in 62e6121 that gets rid of `internal_txn` (by publishing a `pinned_durable_timestamp` separate from the `WT_TXN`-local `durable_timestamp`.  I think it's clearer this way: I was really struggling to convince myself that the previous code was correct...

Reply:
Thanks for the cleanup, @michaelcahill. I think I understand this.
I asked some questions on Slack but I'll record the responses here in case anyone has similar questions:

- The main change is to record the timestamp we use to insert into the durable queue as `pinned_durable_ts`.
- `__wt_txn_clear_read_timestamp` is functionally equivalent and is refactored to make handling of durable and read timestamps as similar as possible.